src/dhtproto/client/legacy/internal/helper/Mirror.d

```diff
@@ -0,0 +1,527 @@
+import dhtproto.client.legacy.internal.helper.mirror.model.MirrorBase;
+
+/*******************************************************************************
+
+    Channel mirror abstract helper class.
+
+    Template params:
+        Dht = type of DHT client (must be derived from DhtClient and contain the
+            RequestScheduler plugin)
+
+*******************************************************************************/
+
+abstract public class Mirror ( Dht : DhtClient ) : MirrorBase!(Dht)
+{
+    import ocean.util.log.Logger;
+    import ocean.text.convert.Formatter;
+    import swarm.Const : NodeItem;
+
+    /// Core functionality shared by Listen and GetAll handlers.
+    private final class SingleNodeRequest
+    {
+        /// Enum defining the state of the request.
+        private enum State
+        {
+            None,
+            Assigned,
+            Scheduled
+        }
+
+        /// State of the request.
+        public State state = State.None;
+
+        /// Logger for this request.
+        private Logger log;
+
+        /// Node that this request is communicating with.
+        private NodeItem node;
+
+        /***********************************************************************
+
+            Constructor.
+
+            Params:
+                id = name / identifier of request, used for logging
+                node = node address/port
+
+        ***********************************************************************/
+
+        public this ( cstring id, NodeItem node )
+        {
+            this.log = Log.lookup(format("Mirror.{} on {}:{} / {}", id,
+                node.Address, node.Port, this.outer.channel_));
+            this.node = node;
+        }
+
+        /***********************************************************************
+
+            Starts the request, if it is not alrady running (or scheduled).
+
+            Params:
+                Rq = type of request
+                rq = request object
+                start_in_ms = milliseconds delay before assigning request. (If
+                    0, assign immediately.)
+
+        ***********************************************************************/
+
+        public void start ( Rq ) ( Rq rq, uint start_in_ms = 0 )
+        {
+            if ( this.state != this.state.None )
+            {
+                this.log.info("Already {}ed; not re{}ing.",
+                    this.state == this.state.Assigned ? "assign" : "schedul",
+                    start_in_ms == 0 ? "assign" : "schedul");
+                return;
+            }
+
+            if ( start_in_ms == 0 )
+            {
+                this.log.info("Starting.");
+                this.outer.dht.assign(rq);
+                this.state = State.Assigned;
+            }
+            else
+            {
+                this.log.info("Starting in {}ms.", start_in_ms);
+                this.outer.dht.schedule(rq, start_in_ms);
+                this.state = State.Scheduled;
```

Comment:
Minor note: this means that one never sees any log output for when a `GetAll` request actually starts, bar the initial one.  One just sees the `Starting in {}ms.` message.